### PR TITLE
support ComponentArrays

### PR DESCRIPTION
**Pull request by scheidan**
Implements #253.
Please let me know if I shoudl change something.

### CODE REVIEW

**Comment by scheidan** [comment]
Some CI tests failed but I think it has nothing to do with this commit. Unfortuantlay it  says only `This check failed` when clicking on details... 

Locally the tests pass.

File: src/hamiltonian.jl [hunk 1]
```diff
@@ -70,15 +70,18 @@ phasepoint(
 
 # If position variable and momentum variable are in different containers,
 # move the momentum variable to that of the position variable.
-# This is needed for AHMC to work with CuArrays (without depending on it).
-phasepoint(
+# This is needed for AHMC to work with CuArrays and other Arrays (without depending on it).
+function phasepoint(
```
Comment (xukai92) [inline src/hamiltonian.jl:74]: Can you implement your change while keeping the current interface of optionally provide `ℓπ` and `ℓκ`?
You can create a function for your version of initializing `r`, e.g. `safe_rsimilar(θ, _r) = (r = similar(θ); copyto!(r, _r); r)`, and update `r=safe_rsimilar((θ, _r))`

**Review by xukai92 — CHANGES_REQUESTED** [review]

**Comment by xukai92** [comment]
Thanks for the PR. 
The proposed way to support ComponentArrays looks great to me.
But the current implementation changes the interface, which is not desirable.
Please check my comments on addressing this.

Also, can you add a test with the use case of ComponentArrays so that we know this PR works as expected?
The example you have in https://github.com/TuringLang/AdvancedHMC.jl/issues/253 should be good enough.
You should be able to add packages only for testing purposes in the `[extras]` section in the Project.toml file.

**Comment by scheidan** [comment]
Thanks for your review @xukai92! Your suggestion made a lot of sense.

**Review by xukai92 — APPROVED** [review]

**Comment by xukai92** [comment]
Great! The changes look good to me. 
It's quite neat to have ComponentArray offically supported in AHMC.
Thanks for the PR again!

**Pull request merged**